contracts/interchain-token-service/src/contract/execute.rs

```diff
@@ -103,20 +121,70 @@ fn ensure_its_source_address(
     Ok(())
 }
 
+fn verify_coin(
+    deps: &DepsMut,
+    coin: &Option<Coin>,
+    message: &Message,
+    source_chain: &ChainNameRaw,
+) -> Result<(), Error> {
+    let expected_coin = gateway_token_transfer(deps, source_chain, message)?;
+    ensure!(
+        &expected_coin == coin,
+        Error::IncorrectAttachedCoin {
+            attached: coin.clone(),
+            expected: expected_coin
+        }
+    );
+    Ok(())
+}
+
+fn gateway_token_transfer(
+    deps: &DepsMut,
+    source_chain: &ChainNameRaw,
+    message: &Message,
+) -> Result<Option<Coin>, Error> {
+    let client: nexus::Client = client::CosmosClient::new(deps.querier).into();
+    let is_core_chain = client
+        .is_chain_registered(&normalize(source_chain))
+        .change_context(Error::NexusQueryError)?;
+
+    if is_core_chain {
+        let token_id = message.token_id();
+        let gateway_denom = state::may_load_gateway_denom(deps.storage, token_id)
+            .change_context(Error::StorageError)?;
+        return match (gateway_denom, message) {
+            (Some(denom), Message::InterchainTransfer { amount, .. }) => Ok(Some(Coin {
+                denom: denom.to_string(),
+                amount: Uint128::try_from(*amount).change_context(Error::TransferAmountOverflow)?,
+            })),
+            _ => Ok(None),
+        };
+    }
+    Ok(None)
```

Comment:
```suggestion
    if !is_core_chain {
        return Ok(None);
    }

    let token_id = message.token_id();
    let gateway_denom = state::may_load_gateway_denom(deps.storage, token_id)
        .change_context(Error::StorageError)?;

    match (gateway_denom, message) {
        (Some(denom), Message::InterchainTransfer { amount, .. }) => Ok(Some(Coin {
            denom: denom.into(),
            amount: Uint128::try_from(*amount).change_context(Error::TransferAmountOverflow)?,
        })),
        _ => Ok(None),
    }
```
return early if not a core chain